src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,90 +183,113 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
-                        FilePath = projectPath,
-                        CompilationOutputAssemblyFilePath = loadedProjectInfo.IntermediateOutputFilePath
-                    };
+                    var (target, targetAlreadyExists) = await GetOrCreateProjectTargetAsync(previousProjectTargets, loadedProjectInfo);
+                    newProjectTargetsBuilder.Add(target);
 
-                    var projectSystemProject = await ProjectFactory.CreateAndAddToWorkspaceAsync(
-                        projectSystemName,
-                        loadedProjectInfo.Language,
-                        projectCreationInfo,
-                        _projectSystemHostInfo);
+                    if (targetAlreadyExists)
+                    {
+                        _ = await target.UpdateWithNewProjectInfoAsync(loadedProjectInfo, hasAllInformation, _logger);
+                    }
+                    else
+                    {
+                        var (targetTelemetryInfo, targetNeedsRestore) = await target.UpdateWithNewProjectInfoAsync(loadedProjectInfo, hasAllInformation, _logger);
+                        needsRestore |= targetNeedsRestore;
+                        telemetryInfos[loadedProjectInfo] = targetTelemetryInfo with { IsSdkStyle = preferredBuildHostKind == BuildHostProcessKind.NetCore };
+                    }
+                }
 
-                    var loadedProject = new LoadedProject(projectSystemProject, ProjectFactory.Workspace.Services.SolutionServices, _fileChangeWatcher, _targetFrameworkManager);
-                    loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(projectToLoad with { ReportTelemetry = false });
-                    existingProjects.Add(loadedProject);
+                var newProjectTargets = newProjectTargetsBuilder.ToImmutableAndFree();
+                foreach (var target in previousProjectTargets)
+                {
+                    // Unload targets which were present in a past design-time build, but absent in the current one.
+                    if (!newProjectTargets.Contains(target))
+                    {
+                        target.Dispose();
+                    }
+                }
 
-                    (targetTelemetryInfo, targetNeedsRestore) = await loadedProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+                if (projectToLoad.ReportTelemetry)
+                {
+                    await _projectLoadTelemetryReporter.ReportProjectLoadTelemetryAsync(telemetryInfos, projectToLoad, cancellationToken);
+                }
 
-                    needsRestore |= targetNeedsRestore;
-                    telemetryInfos[loadedProjectInfo] = targetTelemetryInfo with { IsSdkStyle = preferredBuildHostKind == BuildHostProcessKind.NetCore };
+                if (currentLoadState is ProjectLoadState.Primordial(var projectId))
+                {
+                    // Remove the primordial project now that the design-time build pass is finished.
```

Comment:
```suggestion
                    // Remove the primordial project now that the design-time build pass is finished. This ensures that
                    // we have the new project in place before we remove the primordial project; otherwise for
                    // Miscellaneous Files we could have a case where we'd get another request to create a project
                    // for the project we're currently processing.
```